Sources/SwiftDocC/Infrastructure/DocumentationContext.swift

```diff
@@ -519,77 +519,92 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
-                for doc in documentationNode.docChunks {
-                    let source: URL?
-                    switch doc.source {
-                    case _ where documentationNode.semantic is Article,
-                            .documentationExtension:
-                        source = documentLocationMap[reference]
-                    case .sourceCode(let location, _):
-                        // For symbols, first check if we should reference resolve
-                        // inherited docs or not. If we don't inherit the docs
-                        // we should also skip reference resolving the chunk.
-                        if let semantic = documentationNode.semantic as? Symbol,
-                           let origin = semantic.origin, !externalMetadata.inheritDocs
-                        {
-                            // If the two symbols are coming from different modules,
-                            // regardless if they are in the same bundle
-                            // (for example Foundation and SwiftUI), skip link resolving.
-                            if let originSymbol = documentationCache[origin.identifier]?.semantic as? Symbol,
-                               originSymbol.moduleReference != semantic.moduleReference {
-                                continue
-                            }
-                        }
-
-                        source = location?.url()
-                    }
-
-                    // Find the inheritance parent, if the docs are inherited.
-                    let inheritanceParentReference: ResolvedTopicReference?
-                    if let origin = (documentationNode.semantic as? Symbol)?.origin,
-                       let originReference = documentationCache.reference(symbolID: origin.identifier)
-                    {
-                        inheritanceParentReference = originReference
-                    } else {
-                        inheritanceParentReference = nil
-                    }
-
-                    var resolver = ReferenceResolver(context: self, bundle: bundle, source: source, rootReference: reference, inheritanceParentReference: inheritanceParentReference)
+            guard var documentationNode = try? entity(with: reference),
+                  documentationNode.semantic is Article || documentationNode.semantic is Symbol
+            else {
+                return
+            }
                     
-                    // Update the cache with the resolved node.
-                    // We aggressively release used memory, since we're copying all semantic objects
-                    // on the line below while rewriting nodes with the resolved content.
-                    documentationNode.semantic = autoreleasepool { resolver.visit(documentationNode.semantic) }
+            let inheritanceParentReference: ResolvedTopicReference?
```

Comment:
This closure/function is getting a bit long. Might be nice to extract this portion about inherited documentation into a separate function.

Maybe something like this:

```
let inheritanceParentReference = findSymbolOriginReference(...)

guard externalMetadata.inheritDocs || !docsAreInherited(...) else { return }
```

Reply:
It took a few iterations to find something that I felt was better than leaving the code inline. I ultimately settled on https://github.com/apple/swift-docc/pull/951/commits/55414f2f1681262b4eafbd7220513ba280127ee9 which computed the symbol origin reference inline and passes it to `inheritsDocumentationFromOtherModule(_:symbolOriginReference:)`